lib/bundler/source/mercurial.rb

```diff
@@ -3,8 +3,95 @@
 
 module Bundler
   class Source
-    class Mercurial < Bundler::Plugin::API::Source
-      # Your code goes here...
+    # Class that handled the mercurial source for bundler
+    class Mercurial
+      attr_reader :ref
+
+      def initialize(opts)
+        super
+
+        @ref = options["tag"] || options["branch"] || "default"
+      end
+
+      def api
+        @api ||= Bundler::Plugin::API.new
+      end
+
+      def fetch_gemspec_files
+        @spec_files ||= begin
+          glob = "{,*,*/*}.gemspec"
+          cache_repo unless cached?
+
+          path = if installed?
+            install_path
+          else
+            update_cache changeset
+            cache_path
+          end
+
+          Dir["#{path}/#{glob}"]
+        end
+      end
+
+      def install(spec, opts)
+        api.mkdir_p(install_path.dirname)
+        api.rm_rf(install_path)
+        `hg clone #{cache_path} #{install_path}`
+        api.chdir install_path do
+          `hg update -r #{changeset} 2>&1`
+        end
+
+        nil # No post installation message
+      end
+
+      def optoins_to_lock
+        {
+          "changeset" => changeset,
+          "ref" => ref,
+        }
+      end
+
+    private
+
+      def cache_path
+        @cache_path ||= api.cache_dir.join("soruce-mercurial", repo_name)
+      end
+
+      def cache_repo
+        api.mkdir_p cache_path.dirname
+        `hg clone -U #{uri} #{cache_path} 2>&1`
+      end
+
+      def cached?
+        File.directory?(cache_path)
+      end
+
+      def locked_changeset
+        options["changeset"]
+      end
+
+      def changeset
+        @changeset ||= locked_changeset || latest_changeset
+      end
+
+      def latest_changeset
+        cache_repo unless cached?
+
+        api.chdir(cache_path) do
+          ENV["HGPLAIN"] = "true"
```

Comment:
Yes, I was planing to remove it, but never got around it. Will remove with the next commit.